pkg/test/framework/components/istio/certs.go

```diff
@@ -0,0 +1,157 @@
+//  Copyright Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package istio
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	kubeApiCore "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/test/util/file"
+	"istio.io/istio/pkg/test/util/tmpl"
+)
+
+const (
+	certsConfigTemplate = `
```

Comment:
This feels a bit odd/wrong to me, but I might just not have the context here. What is the user doing when they do this? I thought they install, istio creates the certs, then they copy these?

Reply:
We recommend the user create an offline root, generate intermediate CA signed by that root, and install the intermediate certs in the clusters. Istio signs workload certs with the intermediate. We don't manage the offline root or intermediate CA generation. For testing and demo purposes we've used pre-generated sample certs.

cc @myidpt

Reply:
Do they generate a single intermediate CA shared by all the clusters? Or one single intermediate shared by all clusters?

Just wondering why its an intermediate one here if it is shared - I need to learn more about this topic in general though.

When you install Istio its creating its own root, are we then overriding that? They thing that looked weird to me is I would have thought we set up the cert prior to installing.

Is it possible to do what I described and let Istio created the root cert in one cluster, then copying that to other clusters? I am not suggesting we *should* do that, just checking my understanding

Reply:
+1 ... I could use a better understanding as well.

FYI, the logic here is copied from this [generate.sh](https://github.com/istio/istio/blob/master/tests/testdata/certs/generate.sh) ... I just had to do it myself since the system namespace is configurable.

Reply:
A things:

1) I didn't look closely at the certs below before my previous comment. The cert generation below isn't quite right. `istio.io/istio/sample/certs/Makefile` is a better example. See https://github.com/istio/istio/blob/84423da845508917488d039e53d3096a42d8913d/samples/multicluster/setup-mesh.sh#L97 and https://github.com/istio/istio/blob/84423da845508917488d039e53d3096a42d8913d/samples/multicluster/setup-mesh.sh#L151 for usage. If we do get into the business of automating offline root and intermediate CA automation better tooling (not Makefiles) is required.

2) Revoking a root CA is _much_ more costly than an intermediate CA. The same root (literally) of trust persists in the latter case. 

3) Citadel (and now istiod) use different secrets for self-signed vs. externally plugged. `istio-ca-secret` is the auto-generated self-signed root and it contains the root cert+key. `cacerts` is the external cert and it contains the root cert + intermediate cert/key, and full cert-chain. Both secrets can coexist in the istio-system namespace. Neither is overwritten. Istiod uses cacerts if its available, otherwise it generates istio-ca-secret. This isn't great UX, particularly when upgrading from 1.4 which picked which secret to use based on the --selfSigned flag.

> Is it possible to do what I described and let Istio created the root cert in one cluster.

Technically, yes, though it wouldn't be testing how users configure multicluster.

Reply:
For this PR, I'm fine with creating a root and copy it over to other clusters. That aligns with the multi-cluster demo. Or generating a root cert and plugging it into every cluster.
But we recommend using an offline root, and using the root to sign multiple intermediate certs, one for each cluster to use. This offers many security benefits. I think we should advocate this in the multi-cluster UG. Maybe you can add a TODO here.

Reply:
Intermediate certs is documented in https://istio.io/docs/setup/install/multicluster/gateways/#prerequisites. It's not mentioned at all in the shared control plane setups. The updated multicluster 1.5.1 docs include the intermediate cert recommendation.

Reply:
Thanks for checking, Jason. I think we should use one intermediate cert for each cluster, and align shared control plane setups to do the same.

Reply:
Thanks! that cleared up my confusion

Reply:
@myidpt @ayj @howardjohn PTAL. I'm now generating intermediate certs for each cluster in a similar fashion to the samples Makefile.